optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/uni/UniAnchor.java

```diff
@@ -0,0 +1,184 @@
+            Function<A, GroupKey_> groupKeyMapping) {
+        ConsequenceBuilder._1<?> consequence = on(getAMetadata().getVariableDeclaration())
+                .execute((drools, a) -> {
+                    final A aInlined = inlineA(a);
+                    final GroupKey_ aMapped = groupKeyMapping.apply(aInlined);
+                    RuleContext kcontext = (RuleContext) drools;
+                    kcontext.insertLogical(new LogicalTuple(currentContextId, aMapped));
+                });
+        return Arrays.asList(getAMetadata().getPattern(), consequence);
+    }
+
+    public List<RuleItemBuilder<?>> terminateWithScoring(Global<? extends AbstractScoreHolder> scoreHolderGlobal) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, __) -> {
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext);
+        });
+    }
+
+    public List<RuleItemBuilder<?>> terminateWithScoring(Global<? extends AbstractScoreHolder> scoreHolderGlobal,
+            ToIntFunction matchWeighter) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, a) -> {
+            int weightMultiplier = matchWeighter.applyAsInt(inlineA(a));
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext, weightMultiplier);
+        });
+    }
+
+    public List<RuleItemBuilder<?>> terminateWithScoring(Global<? extends AbstractScoreHolder> scoreHolderGlobal,
+            ToLongFunction matchWeighter) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, a) -> {
+            long weightMultiplier = matchWeighter.applyAsLong(inlineA(a));
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext, weightMultiplier);
+        });
+    }
+
+    public <A> List<RuleItemBuilder<?>> terminateWithScoring(Global<? extends AbstractScoreHolder> scoreHolderGlobal,
+            Function<A, BigDecimal> matchWeighter) {
+        return terminateWithScoring(scoreHolderGlobal, (drools, scoreHolder, a) -> {
+            BigDecimal weightMultiplier = matchWeighter.apply(inlineA(a));
+            RuleContext kcontext = (RuleContext) drools;
+            scoreHolder.impactScore(kcontext, weightMultiplier);
+        });
+    }
+
+    private <A, ScoreHolder extends AbstractScoreHolder> List<RuleItemBuilder<?>> terminateWithScoring(
+            Global<ScoreHolder> scoreHolderGlobal, Block3<Drools, ScoreHolder, A> consequenceImpl) {
+        ConsequenceBuilder._2<ScoreHolder, A> consequence =
+                on(scoreHolderGlobal, (Declaration<A>) getAMetadata().getVariableDeclaration())
+                        .execute(consequenceImpl);
+        return Arrays.asList(getAMetadata().getPattern(), consequence);
+    }
+
+    public static String createContextId() {
+        return UUID.randomUUID().toString();
```

Comment:
id must be reproducible between runs -> count++

Reply:
rule name == the contextId of the inserted logical tuple of the same rule that inserts it